### DISCUSSION TITLE
Do compare functions works correctly?

### DISCUSSION THREAD

**Discussion by prokopiy** [lfe/lfe#521] [Q&A]
In LFE:
lfe> (> 2 1)
true
lfe> (> 3 2 1)
true
lfe> (> 1 2 1)
true
lfe> (< 1 2 3)
false

In CLISP:
[1]> (> 2 1)
T
[2]> (> 3 2 1)
T
[3]> (> 1 2 1)
NIL
[4]> (< 1 2 3)
T

**Comment by oubiwann** [comment]
I think I may have run into that in my early days of LFE; might have asked about it on the mail list? Anyway, I'm pretty sure the only way to do this without a macro is via `and`:

```lisp
lfe> (and (> 1 2) (> 2 1))
false
```

The same is true in Erlang itself. Compare:

```erlang
1> 1 > 2 > 1.
* 1: syntax error before: '>'
```

with this:

```erlang
1> (1 > 2) and (2 > 1).
false
```

I'll take a look at the source and see what Robert's intentions might have been. (I suspect, though, it was to keep it simple and keep it close to Erlang).

**Comment by oubiwann** [comment]
So it looks like 2-arity comparisons are handled explicitly with a call to `exp_bif`; for 3+arity it's a little more involved. There's a fold happening over the arguments, with the accumulator and head of the remaining args being compared, which leads me to believe that only if booleans are passed in the chain would this work [edit: "work as expected"]; it seems that anything else wouldn't [again: "wouldn't work as expected"].

Yeah, check this out:

```lisp
lfe> (< 'true 'false)
false
lfe> (> 'true 'false)
true
lfe> (> 'true 'false 'false)
true
lfe> (> 'true 'false 'false 'false)
true
lfe> (> 'true 'false 'false 'true)
false
```

The next question is: "Is this expected? Ideal?" etc. The solution will (obv) be more complex if changed to support all types ...

**Comment by oubiwann** [comment]
Oh, btw, I was looking at `exp_comp/3` in `src/lfe_macro.erl`.

**Comment by oubiwann** [comment]
So, coming back to your question in the ticket ... I'd say provisionally "yes" (they do work as expected). It's just not what non-Erlang-knowing Lispers might expect.

I'm sure we could parley this into a feature request, though. The current `foldl` won't cut it.  I think what would basically have to happen is that instead of doing incremental / iterative evaluation for the n-arity case (where `n` is greater than 2), we'd have to convert the args to a set of 2-arity comparison expressions -- and then evaluate _that_. 

Maybe there's a better way, though?

**Comment by yurrriq** [comment]
If we agree to support variadic comparison operators, I'd volunteer to implement it :smile: Off the top of my head, I can't think of a better approach than translating:
```lfe
(> x1 x2 x3 ... xm xn)
```
into, effectively:
```lfe
(and (> x1 x2) (> x2 x3) ... (> xm xn))
```
Would we then allow polyadic comparisons in guards too?

**Comment by rvirding** [comment]
The comparison operators are now implemented as macros and I see that they are probably not all correct. My plan was that even though they are implemented as macros they should behave as functions and evaluate all their arguments. They were modelled on the CL comparison operators. With a little bit of cunning they should be made guard safe.

A problem is expanding it in such a way that each element is only evaluated once. That could be a problem in guards where you are not allowed to bind variables.
